### DISCUSSION TITLE
Compiling for GameCube?

### DISCUSSION THREAD

**Discussion by shazz** [HTV04/wiilove#8] [General]
Hi,

I tried to recompile wiilove for the gamecube but there is a dependency on `luajit`, where can I find this library?

Thanks!

**Comment by shazz** [comment]
Ah I guess this is this one https://github.com/HTV04/LuaJIT :)

**Comment by htv04** [comment]
Hey, yeah that would be the one 👍

Beware that I haven't added GameCube support for LuaJIT yet (mostly because I haven't had an opportunity to test it), so you'll need to tinker with the code a little to get it to compile.

Also it's worth noting that GameCube support is definitely coming to WiiLÖVE. I even hinted at it in some source files. However, any help getting GameCube support ready would be very appreciated :)

**Comment by shazz** [comment]
Yeah I did recompile for the GC LuaJIT, grrlib-mod, png-mod and finally Wiilove... sorry Cubelove :) but got a black screen when running the wiilove.dol (but I did not know what to expect)

I'll go back and check first the deps I recompiled still work :)

Happy to help !

**Comment by htv04** [comment]
Just wondering, are you testing this on real hardware, and if so, do you have an SD Gecko? Unfortunately Dolphin can't emulate those at the moment, but if you have one, in theory you should be able to store and load WiiLÖVE games from it.

**Comment by shazz** [comment]
Yeah I'm only testing on a real GameCube, I found Dolphin pretty...inaccurate, most of my stuff just doesn't work in Dolphin.

To debug I have a USB Gecko but I prefer to use the BBA, faster and more reliable, I wrote a simple network launcher / debugger (similar to PS2Link).

I also have an SD Gecko and a sd2sp2, I'll check how you're mounting the fat, normally easier with libogc2 + libfat than libogc.

So I tested luaJIT with a simple API example, it works on my GC, so I guess I broke something in Wiilove.

**Comment by htv04** [comment]
I've heard of libogc2, it seems to have a lot of fixes libogc doesn't have. However, it also doesn't include some commits from libogc. Do you recommend that I switch over to libogc2 for WiiLÖVE?

**Reply by shazz** [reply]
Frankly speaking, I asked myself the same question (and I'm asking it too regarding grrlib-mod as I found some bugs in grrlib and not sure crayon will fix them, another discussion!)

But at the end, libogc2 is fairly more active and maintained by I guess the most knowledgeable developer (Swiss is just a crazy piece of software) on the GC, extrems. And I needed his modifications on the FS and on the exception vectors. And I'm pretty sure he's looking at the libogc commits.

And in fact it costs nothing, I did not see any breaking changes. So I finally switched and only use libogc2 and happy with it.

My 2 cents :)

**Reply by htv04** [reply]
Hmm, I guess I'll look into using it then! Hopefully it isn't complicated to set up.

> I asked myself the same question (and I'm asking it too regarding grrlib-mod as I found some bugs in grrlib and not sure crayon will fix them, another discussion!)

GRRLIB-mod has a ton of changes that set it apart from normal GRRLIB. I added a set of 2D matrix functions, so you can translate, scale, and rotate the drawing matrix. I also added a few more settings. The most recent change I made was creating a new GRRLIB_texture object, intended to replace GRRLIB_texImg, that is way more optimized. They have a GXTexObj stored instead of having one be unnecessarily generated every frame (and changes can be made to the data before "finalizing" it, generating the GXTexObj). Oh, and TPL files are now supported!

However, all of these changes make GRRLIB-mod API-incompatible with the original GRRLIB, so you would need to change existing GRRLIB projects' code in order to take advantage of GRRLIB-mod's features.

**Reply by shazz** [reply]
No, using libogc2 is really straight forward, simply is this kind of makefile to use it instead of libogc: https://github.com/PacMooseIncorporated/GameCube4Fun/blob/main/libogc2/console/Makefile

**Reply by shazz** [reply]
I took a look at your additions, really makes sense! And also at your unanswered pull request. 
I think I will use yours now and contribute if you're ok.

Maybe it could be called grrlib2 ? 😂

**Comment by shazz** [comment]
So to go back to WiiLÖVE for the Gamecube:
- nothing to change in grrlib-mod, just to be sure the `$PLATFORM` is set to `cube`
- for LuaJIT, not that much to change
   - in `make.sh`, set `TARGET_SYS=cube`
   - in `install.sh`, copy headers in `$DEVKITPRO/portlibs/gamecube/include` and libs in `DEVKITPRO/portlibs/gamecube/lib`. That's it!

I coded this small test program to check it works: https://github.com/PacMooseIncorporated/GameCube4Fun/tree/main/libogc2/lualib

**Reply by htv04** [reply]
Got it, when I get home I'll fix up LuaJIT for the GameCube.

Also I tried using libogc2, but I was met with a sea of warnings. When I booted WiiLÖVE, I was greeted by a black screen. Maybe this is the black screen you were getting?

It seems that libogc2 split from libogc sometime in 2020. Extrems ported over some commits since then, but missed a lot of them. However, I really like some of the changes libogc2 has. So I decided to take it upon myself to port over a lot of these commits as individual PRs: https://github.com/extremscorner/libogc2/pulls?q=is%3Apr+author%3AHTV04

These commits combined fix the warnings and add some missing features, but I still need to figure out what's causing the black screen, since I haven't tested the fixed libogc2 with WiiLÖVE yet.

**Reply by shazz** [reply]
In fact to test WiiLÖVE I did not compiled using `libogc2` but only latest `libogc`. 

Can I ask you 2 little favors? 
1. Could you write a nice C++ wrapper to my [libgctools](https://github.com/PacMooseIncorporated/GameCube4Fun/tree/main/libgctools) library that I Could use (that's basically a BBA remote `printf` :D), I'd like to avoid polluting your well designed C++ code by some ugly C calls. Then I'll add some calls to check where it get stuck (maybe `boot.lua` but not sure)
2. *[less important right now]* Can you `#ifdef` all references to the wiimotes/WPAD (this is basically what I disabled from your code to test)

That's pretty cool you proposed those PRs, I hope extrems will take them into account!

**Reply by htv04** [reply]
Interesting. Perhaps it's some random function from libogc that only works on the Wii. Or maybe GRRLIB-mod?

1. I could try, but are you sure that's necessary? It would be good for looks, but not any practical use (if we're talking internal debugging and not public code). libogc and GRRLIB-mod also pollute the global namespace already, so libgctools wouldn't necessarily be out of place.
2. Of course! I planned to add `#ifdef` statements when I started working on GameCube support, but I wanted to focus on Wii support before then, lol.

**Reply by shazz** [reply]
I'll continue to check...

1. For my `libgctools ` debug library, yeah, don't add it the master, only in a debug branch or whatever the time for me to debug and find the problem, then we can get rid of it :) And you'll be better than me to avoid polluting too much :)

1. Cool!

**Reply by htv04** [reply]
> For my `libgctools `debug library, yeah, don't add it the master, only in a debug branch or whatever the time for me to debug and find the problem, then we can get rid of it :) And you'll be better than me to avoid polluting too much :)

I think you might have misunderstood what I said, I know how libgctools can be used to find the pinpoint the issue on the GameCube, but I was referring to writing a C++ wrapper for it. If we're just talking about debugging WiiLÖVE, I think making a wrapper would only be good for looks. Which would be cool, but not really practical.

I think I might be misunderstanding your post too, but do you want me to set up a branch with libgctools linked so you can debug it? I can set that up if you want.

**Reply by shazz** [reply]
That's fine, my question was only what is the easiest way to integrate `libgctools` within your code even for a limited time :)

**Reply by htv04** [reply]
I added HW_DOL macros to WiiLÖVE's code, so it should compile and run on GameCube without issues. Are you able to integrate `libgctools` or do you still want me to set up a branch with it?

**Reply by htv04** [reply]
Decided to make a fork of libogc, featuring my AESND changes, some fixes and enhancements, and perhaps most importantly, libogc2 merged in :)

https://github.com/HTV04/libogc-mod

**Reply by htv04** [reply]
After a lot of trial and error, it seems that `fatInitDefault()` is broken, causing a black screen whenever it is run. Going to create a libogc2 issue about this because I can't seem to find out why this is happening.
